### Codio Activity 4.3: Data Cleaning

**Expected Time: 120 Minutes**

**Total Points: 40**

This assignment focuses on making decisions with a somewhat dirty dataset.  From datatype conversion, to object removal and cleaning, to missing value handling, cleaning data will be a regular activity while working with data in the wild.  The dataset comes from kaggle [here](https://www.kaggle.com/shansh8/companies-info-wikipedia-2021), and represents information on businesses pulled from [dbpedia](http://dbpedia.org) pertaining to business indicators from a variety of international companies.


#### Index:

- [Problem 1](#Problem-1)
- [Problem 2](#Problem-2)
- [Problem 3](#Problem-3)
- [Problem 4](#Problem-4)
- [Problem 5](#Problem-5)
- [Problem 6](#Problem-6)
- [Problem 7](#Problem-7)
- [Problem 8](#Problem-8)
- [Problem 9](#Problem-9)


![](images/dbpedia.png)

### Load and Explore the Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [2]:
df = pd.read_csv('data/biz.zip', compression = 'zip')

In [3]:
df.head()

a  \
0      http://dbpedia.org/resource/Deutsche_EuroShop   
1      http://dbpedia.org/resource/Deutsche_EuroShop   
2  http://dbpedia.org/resource/Industry_of_Machin...   
3  http://dbpedia.org/resource/Industry_of_Machin...   
4  http://dbpedia.org/resource/Industry_of_Machin...   

                                 name  \
0                Deutsche EuroShop AG   
1                Deutsche EuroShop AG   
2  Industry of Machinery and Tractors   
3  Industry of Machinery and Tractors   
4  Industry of Machinery and Tractors   

                                                type foundation  \
0     http://dbpedia.org/resource/Aktiengesellschaft       1999   
1     http://dbpedia.org/resource/Aktiengesellschaft       1999   
2  http://dbpedia.org/resource/Društvo_sa_ogranič...  8/12/1996   
3  http://dbpedia.org/resource/Društvo_sa_ogranič...  8/12/1996   
4  http://dbpedia.org/resource/Društvo_sa_ogranič...  8/12/1996   

                               location  \
0   http://dbpedia.org/resource/Germany   
1   http://dbpedia.org/resource/Hamburg   
2  http://dbpedia.org/resource/Belgrade   
3    http://dbpedia.org/resource/Serbia   
4  http://dbpedia.org/resource/Belgrade   

                                            products  \
0                     Investment in shopping centers   
1                     Investment in shopping centers   
2               http://dbpedia.org/resource/Tractors   
3  http://dbpedia.org/resource/Agricultural_machi...   
4  http://dbpedia.org/resource/Agricultural_machi...   

                                      sector  employee    assets    equity  \
0  http://dbpedia.org/resource/Real_property         4  2.96E+09  1.25E+09   
1  http://dbpedia.org/resource/Real_property         4  2.96E+09  1.25E+09   
2   http://dbpedia.org/resource/Agribusiness         4  4.65E+07         0   
3   http://dbpedia.org/resource/Agribusiness         4  4.65E+07         0   
4   http://dbpedia.org/resource/Agribusiness         4  4.65E+07         0   

    revenue          profit  
0  2.24E+08        8.04E+07  
1  2.24E+08        8.04E+07  
2     30000  -€0.47 million  
3     30000  -€0.47 million  
4     30000  -€0.47 million

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   a           10000 non-null  object
 1   name        9969 non-null   object
 2   type        10000 non-null  object
 3   foundation  9665 non-null   object
 4   location    9936 non-null   object
 5   products    9014 non-null   object
 6   sector      10000 non-null  object
 7   employee    10000 non-null  int64 
 8   assets      7103 non-null   object
 9   equity      7124 non-null   object
 10  revenue     6943 non-null   object
 11  profit      7107 non-null   object
dtypes: int64(1), object(11)
memory usage: 937.6+ KB


[Back to top](#Index:) 

### Problem 1

#### Plot of Missing Data

**0 Points**

Create a barplot using `matplotlib` with the $x$-axis representing the columns and the $y$-axis representing the count of missing values. Sort these values from least to greatest.  Save your plot in the `results` folder as `missing_plot.png`.  

In [5]:
### GRADED


# YOUR CODE HERE
df.isnull().sum().sort_values().plot(kind = 'bar')
plt.savefig('results/missing_plot.png')
plt.close()

[Back to top](#Index:) 

### Problem 2

#### `name` column

**5 Points**

The column `name` has 31 missing values.

Subset the data to examine these observations where `name` is missing. Assign the result as a DataFrame to `ans2` below.  

In [6]:
### GRADED

ans2 = None

# YOUR CODE HERE
ans2 = df[df['name'].isnull()].copy()

# Answer check
print(type(ans2))
ans2.head()

<class 'pandas.core.frame.DataFrame'>


a name  \
3906  http://dbpedia.org/resource/Royal_Canadian_Mint  NaN   
3909  http://dbpedia.org/resource/Royal_Canadian_Mint  NaN   
4861   http://dbpedia.org/resource/Bulgarian_Postbank  NaN   
4862   http://dbpedia.org/resource/Bulgarian_Postbank  NaN   
4863   http://dbpedia.org/resource/Bulgarian_Postbank  NaN   

                                                   type foundation  \
3906  http://dbpedia.org/resource/Crown_corporations...   1/2/1908   
3909  http://dbpedia.org/resource/Crown_corporations...   1/2/1908   
4861        http://dbpedia.org/resource/Commercial_bank       1991   
4862        http://dbpedia.org/resource/Commercial_bank       1991   
4863        http://dbpedia.org/resource/Commercial_bank       1991   

                                  location  \
3906                                   520   
3909               Ottawa, Ontario, Canada   
4861  http://dbpedia.org/resource/Bulgaria   
4862  http://dbpedia.org/resource/Bulgaria   
4863  http://dbpedia.org/resource/Bulgaria   

                                             products  \
3906                http://dbpedia.org/resource/Coins   
3909                http://dbpedia.org/resource/Coins   
4861     http://dbpedia.org/resource/Asset_management   
4862  http://dbpedia.org/resource/Factoring_(finance)   
4863   http://dbpedia.org/resource/Investment_banking   

                                              sector  employee    assets  \
3906     http://dbpedia.org/resource/Mint_(facility)      1280  4.44E+08   
3909     http://dbpedia.org/resource/Mint_(facility)      1280  4.44E+08   
4861  http://dbpedia.org/resource/Financial_services      3000  6.82E+09   
4862  http://dbpedia.org/resource/Financial_services      3000  6.82E+09   
4863  http://dbpedia.org/resource/Financial_services      3000  6.82E+09   

        equity          revenue    profit  
3906  1.85E+08  $2,641.4million  2.45E+07  
3909  1.85E+08  $2,641.4million  2.45E+07  
4861  1.15E+09         3.05E+08  1.20E+08  
4862  1.15E+09         3.05E+08  1.20E+08  
4863  1.15E+09         3.05E+08  1.20E+08

[Back to top](#Index:) 

### Problem 3

### Filling Missing Values in `name`

**5 Points**

Examining the rows missing values in names shows that in the adjacent `a` column, the end of the url contains what can stand in as a name.  Using the `split` string method on the column `a` to split the URL by `/`.  

Assign your split url's as a series to `ans3` below.

This series will be a collection of lists:

```python
0       [http:, , dbpedia.org, resource, Deutsche_Euro...
1       [http:, , dbpedia.org, resource, Deutsche_Euro...
2       [http:, , dbpedia.org, resource, Industry_of_M...
3       [http:, , dbpedia.org, resource, Industry_of_M...
4       [http:, , dbpedia.org, resource, Industry_of_M...
         .
         .
         .
```



In [7]:
### GRADED

ans3 = None

# YOUR CODE HERE
ans3 = df['a'].str.split('/')

# Answer check
print(type(ans3))
ans3.head()

<class 'pandas.core.series.Series'>


0    [http:, , dbpedia.org, resource, Deutsche_Euro...
1    [http:, , dbpedia.org, resource, Deutsche_Euro...
2    [http:, , dbpedia.org, resource, Industry_of_M...
3    [http:, , dbpedia.org, resource, Industry_of_M...
4    [http:, , dbpedia.org, resource, Industry_of_M...
Name: a, dtype: object

[Back to top](#Index:) 

### Problem 4

#### Extracting the names

**5 Points**

To extract the last element of this list, use the `.apply` method together with an appropriate `lambda` function to create a series based on the last entry of the above list in each row.  Your results should begin with:

```python
0                        Deutsche_EuroShop
1                        Deutsche_EuroShop
2       Industry_of_Machinery_and_Tractors
3       Industry_of_Machinery_and_Tractors
4       Industry_of_Machinery_and_Tractors
                       ...                
```

Save this series to `ans4` below.

In [8]:
### GRADED

ans4 = None

# YOUR CODE HERE
ans4 = df['a'].str.split('/').apply(lambda x: x[-1])

# Answer check
print(type(ans4))
ans4.head()

<class 'pandas.core.series.Series'>


0                     Deutsche_EuroShop
1                     Deutsche_EuroShop
2    Industry_of_Machinery_and_Tractors
3    Industry_of_Machinery_and_Tractors
4    Industry_of_Machinery_and_Tractors
Name: a, dtype: object

[Back to top](#Index:) 

### Problem 5

#### Filling the Missing Values in `name` 

**5 Points**

The `.fillna` method can accept a series and fill in the missing values based on the matching indicies.  


Use the series `ans4` you created in Problem 4 to fill in the missing values in the `name` column. 

Overwrite the earlier column and create a DataFrame with no missing values in the `name` column and assign this new DataFrame to `ans5` below.

In [9]:
### GRADED

ans5 = None

# YOUR CODE HERE
ans5 = df.copy()
ans5['name'] = ans5['name'].fillna(ans4)

# Answer check
print(type(ans5))
ans5['name'].isnull().sum()

<class 'pandas.core.frame.DataFrame'>


0

[Back to top](#Index:) 

### Problem 6

#### `location` missing values

**5 Points**

Note that the `location` column is still missing data.  The column is supposed to represent the location of the company as a URL in `dbpedia` or in some cases simply the name of the city the company is located in.  


Use the `isnull()` function to count how many missing values ae in the `locatio` columns. Next, use the `value_counts()` function on the `foundation` column.  Assign these counts as a series to `ans6` below.

In [10]:
### GRADED

ans6 = None

# YOUR CODE HERE
ans6 = df[df['location'].isnull()].foundation.value_counts()

# Answer check
print(type(ans6))
ans6.head()

<class 'pandas.core.series.Series'>


4/8/1980                        34
Colombo, Sri Lanka              12
9/7/1933                         6
New Orleans, Louisiana, U.S.     3
as Visioneer                     2
Name: foundation, dtype: int64

[Back to top](#Index:) 

### Problem 7

#### Replace missing `location` with `foundation`

**5 Points**

While not perfect, the values in the `foundation` column could serve as a fill in for the missing values in `location`.  Replace the missing values in `location` with their corresponding value in `foundation`.  Assign the series with the value filled for `location` to `ans7` below.

In [11]:
### GRADED

ans7 = None

# YOUR CODE HERE
ans7 = ans5['location'].copy()
ans7 = ans7.fillna(ans5['foundation'])

# Answer check
print(type(ans7))
ans7

<class 'pandas.core.series.Series'>


0        http://dbpedia.org/resource/Germany
1        http://dbpedia.org/resource/Hamburg
2       http://dbpedia.org/resource/Belgrade
3         http://dbpedia.org/resource/Serbia
4       http://dbpedia.org/resource/Belgrade
                        ...                 
9995                     London, England, UK
9996                     London, England, UK
9997                     London, England, UK
9998                     London, England, UK
9999                     London, England, UK
Name: location, Length: 10000, dtype: object

[Back to top](#Index:) 

### Problem 8

#### Drop rows missing revenue and profit

**5 Points**


Note that the `revenue` and `profit` columns do not have all their values.  


Use the function `dropna()` on `df` to select the non-null entries in the `subset` with columns `revenue` and `profit`.  

Assign you answer as a DataFrame to `ans8` below.

In [12]:
### GRADED

ans8 = None

# YOUR CODE HERE
df['name'] = df['name'].fillna(df['a'].str.split('/').apply(lambda x: x[-1]))
ans8 = df.dropna(subset = ['revenue', 'profit'])

# Answer check
print(type(ans8))
ans8.shape

<class 'pandas.core.frame.DataFrame'>


(5204, 12)

[Back to top](#Index:) 

### Problem 9

#### How many rows were lost?

**5 Points**

Compare the shape of the original dataset to your solution in 8.  How many rows were lost dropping the data?  Assign your answer as an integer to `ans9` below.

In [13]:
### GRADED

ans9 = None

# YOUR CODE HERE
original_rows = df.shape[0]
remaining_rows = ans8.shape[0]
ans9 = original_rows - remaining_rows

# Answer check
print(type(ans9))
ans9

<class 'int'>


4796

Note that you've lost almost half of our data dropping missing revenue and profit.  Perhaps a different approach should be considered!  Take some time and examine other missing data in our dataset and some strategies for dealing with it.  If it is possible to keep observations do so.

### Summary of Exercises

This notebook focuses on data cleaning techniques using a business dataset from DBpedia. The exercises cover:

1. **Missing Data Visualization** - Creating a bar plot to visualize missing values across columns
2. **Handling Missing Names** - Identifying and examining rows with missing company names (31 missing values)
3. **URL Parsing** - Splitting URLs to extract company names from the 'a' column
4. **String Manipulation** - Using lambda functions to extract the last element from split URLs
5. **Value Imputation** - Filling missing company names using extracted values from URLs
6. **Location Data Analysis** - Examining missing location data and its relationship with the foundation column
7. **Location Data Imputation** - Using foundation data to fill missing location values
8. **Row Filtering** - Removing rows with missing revenue and profit data
9. **Data Loss Assessment** - Quantifying the impact of dropping rows (4,796 rows lost)

### Key Takeaways

1. **Data Quality Assessment**
   - Visualizing missing data helps understand the scope of data cleaning needed
   - Multiple columns contained missing values requiring different cleaning approaches

2. **Creative Data Recovery**
   - URLs contained useful information that could be extracted to fill missing company names
   - Related columns (foundation) could serve as alternative sources for missing data (location)

3. **Trade-offs in Data Cleaning**
   - Dropping rows with missing revenue/profit resulted in losing almost half the dataset
   - This highlights the importance of considering alternative approaches to handle missing data
   - Balance needed between data completeness and quality

4. **Technical Skills Applied**
   - String manipulation (URL parsing)
   - pandas operations (fillna, dropna, value_counts)
   - Lambda functions
   - Data visualization